pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -201,6 +236,62 @@ func (e *EndpointController) addPod(obj interface{}) {
 	}
 }
 
+func (e *EndpointController) addNode(obj interface{}) {
+	node := obj.(*v1.Node)
+	if err := e.nodePodsCache.AddNode(node); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Unable to add node %v to nodePodsCache: %v", node.Name, err))
+	}
+}
+
+func (e *EndpointController) deleteNode(obj interface{}) {
+	if node, ok := obj.(*v1.Node); ok {
+		if err := e.nodePodsCache.RemoveNode(node); err != nil {
+			utilruntime.HandleError(fmt.Errorf("Unable to remove node %v from nodePodsCache: %v", node.Name, err))
+		}
+	}
+
+	// If we reached here it means the node was deleted but its final state is unrecorded.
+	tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get object from tombstone %#v", obj))
+		return
+	}
+	node, ok := tombstone.Obj.(*v1.Node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a Node: %#v", obj))
+		return
+	}
+	if err := e.nodePodsCache.RemoveNode(node); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Unable to remove node %v from nodePodsCache: %v", node.Name, err))
+	}
+}
+
+func (e *EndpointController) updateNode(old, cur interface{}) {
+	// We only need to update endpoint address when node labels changed.
+	oldNode := old.(*v1.Node)
+	curNode := cur.(*v1.Node)
+	if apiequality.Semantic.DeepEqual(oldNode.Labels, curNode.Labels) {
+		return
+	}
+	pods, err := e.nodePodsCache.ListPodsOnNode(curNode.Name)
```

Comment:
`pods` maybe data race.

Reply:
Please check `ListPodsOnNode()` in cache.go

```
func (cache *NodePodsCache) ListPodsOnNode(node string) ([]*v1.Pod, error) {
	cache.mu.Lock()
	defer cache.mu.Unlock()

	return cache.NodePods[node], nil
}
```

We protect the `nodePods` cache via mutex.